repair/row_level.cc

```diff
@@ -650,18 +653,59 @@ future<std::list<repair_row>> to_repair_rows_list(repair_rows_on_wire rows, sche
     co_return std::move(row_list);
 }
 
+std::unordered_set<dht::token> get_tokens_from_repair_row_list(const std::list<repair_row>& rows) {
+    std::unordered_set<dht::token> ret;
+    for (auto& row : rows) {
+        // Get the tokens for the rows that are read from the node itself and
+        // ignore the rows received from peer nodes. A row with dirty_on_master
+        // flag set comes from peer node. See to_repair_rows_list.
+        if (!row.dirty_on_master()) {
+            ret.insert(row.get_dk_with_hash()->dk.token());
+        }
+    }
+    return ret;
+}
+
+// Check if the row is expired and it does not shadow any data read from the local node.
+bool can_skip_expired_mf(schema_ptr s, const dht::token& t, mutation_fragment& mf,
+        const std::list<repair_row>& rows,
+        std::optional<std::unordered_set<dht::token>>& keys_from_local_node) {
+    bool skip_write_mf = false;
+    if (mf.is_clustering_row()) {
+        auto& cr = mf.as_clustering_row();
+        auto live = cr.is_live(*s, tombstone(), gc_clock::now());
+        if (!live) {
+            if (!keys_from_local_node) {
+                keys_from_local_node = get_tokens_from_repair_row_list(rows);
```

Comment:
Shouldn't we make the check directly from rows?

If the problem is the lookup, we can advance a cursor along rows.

Reply:
We could walk though the list of rows and do the check for each fragment. But it is too slow.

Reply:
To balance check accuracy and check overhead, I intentionally used the granularity of partitions and not rows to do the check. 
We could have the case: a row covers no local data but the check think it covers data, so we do not apply the skip optimization but it is safe in correctness. 

Later we could do a two stage checks, first at the partition level, if hit, check the row level. It is more complicated but more accurate. 

For now, simplicity and correctness are more important than the accuracy.

Reply:
I don't understand the motivation for this PR.
It tries to optimise a corner case, but it is partial and also simple to the point that allows common false positive. Not only there is a check only for the existsnce of the whole token, but there is no check for overlap at the write timestamp range, so the conditions allowing the optimization seem overly narrow to me to justify the optimization.

The main advantage that row level repair has over compaction in the area of tombstone purging is the visibility of rows, so unless we can make use of it, I don't see the point of doing that in repair.

Reply:
Benny, check the example in the commit message. That's the motivation for this
PR. I do not think this is a corner case. This happens in the field and hurts
users. Checking at the token or row granularity is just an implementation
detail. The same goes for the writing timestamp of the row (e.g., A expired row
r1 covers a local row r2 in terms of keys, but r2 has a newer timestamp).

Checking with row granularity is making use of the visibility of rows, we can
definitely have it.

The first priority of this patch is to explore if the mechanism works and if it
is safe, but not covering all the possible optimization we could ever have.

Reply:
Good to know that we are on the same page of the "ping-pong" scenario ;-)

Again, I want to mention we have no problem to check at the row granularity. 

Skipping to write expired rows during repiar is not doing the job of compaction to purge expired data. It is just not syncing them. It is the job of repair to decide what to sync and what not to sync. 

In theory, it is best if we could avoid writing them to disk than replying compaction to get rid of them later after the writes. The saving is obvious:  cpu and disk usage to write, cpu and disk usage to purge them. 

Anyway, we need to handle this very carefully.  I am open to different solutions. Let's think together.